lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -203,14 +231,40 @@ OpFoldResult Aten__Is__Op::fold(ArrayRef<Attribute> operands) {
 // AtenLenTOp
 //===----------------------------------------------------------------------===//
 
+OpFoldResult AtenDimOp::fold(ArrayRef<Attribute> operands) {
+  if (auto tensorType = getOperand().getType().dyn_cast<BaseTensorType>()) {
+    if (tensorType.hasSizes())
+      return IntegerAttr::get(IntegerType::get(getContext(), 64),
+                              tensorType.getSizes().size());
+  }
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+// AtenLenTOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult AtenLenTOp::fold(ArrayRef<Attribute> operands) {
+  // `len([1,1,1])` -> `3`
+  if (auto buildList = getOperand().getDefiningOp<Basicpy::BuildListOp>()) {
+    return IntegerAttr::get(IntegerType::get(getContext(), 64),
+                            buildList.getNumOperands());
+  }
+  return nullptr;
+}
+
 void AtenLenTOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
                                              MLIRContext *context) {
+  // `len(t.size())` -> `t.ndim`
   patterns.add(+[](AtenLenTOp op, PatternRewriter &rewriter) {
-    auto buildList = op.getOperand().getDefiningOp<Basicpy::BuildListOp>();
-    if (!buildList)
-      return rewriter.notifyMatchFailure(op, "operand not basicpy.build_list");
-    rewriter.replaceOpWithNewOp<::mlir::ConstantOp>(
-        op, rewriter.getI64IntegerAttr(buildList.getNumOperands()));
+    auto size = op.getOperand().getDefiningOp<AtenSizeOp>();
+    if (!size)
+      return rewriter.notifyMatchFailure(op, "operand not AtenSizeOp");
+    // TODO: Normalize all the torch scalar integer types to consistently use
+    // one type (probably si64 or a new `!torch.int`) so that this op and others
```

Comment:
Probably a `!torch.int`. There's already enough of a semantic gap between Python ints (which tend to be arbitrary magnitude) and Torch/et-al ints (fixed bit depth), it would be nice to preserve the fact that we are working on a `!torch.int` and not just a thing that was prematurely pinned to an `i64`